Implement next/head (Open Graph, Twitter, locale alternates, Google Analytics)

lennerd opened the pull request:
This PR uses next/head to add different important meta and link HTML elements to the head of the page:

- Open Graph
- Twitter
- Locale Alternates
- Google Analytics

Open tasks:
- [x] Add images (uses a fallback image for now)
- [x] Decision on where to get Twitter handle and GA tracking ID
- [x] Tracking ID for Google Analytics 
- [x] Enable switching locale via query parameter
- [ ] Store locale from query parameter in cookie
- [x] Set lang attribute in html element

lennerd commented:
Rebased on current feat/ssr.

mutaphysis commented on src/AsyncNextHead.js:
```diff
@@ -0,0 +1,66 @@
+
+  handlePromiseFetched(headPromise: Promise<React$Element<any>>, head: React$Element<any>) {
+    if (this.state.headPromise !== headPromise) {
+      return;
```
+1: Good one!

mutaphysis commented on src/AsyncNextHead.js:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  render() {
+    return <Head>{this.state.head}</Head>;
```
0: Indicate loading in head?

lennerd replied:
Not sure what you mean? We won't be able to show a loading indicator here, are we?

mutaphysis replied:
Instead of rendering nothing when the promise is not yet resolved, we could render a loading title. Not really needed, though.

lennerd replied:
Ah, I see. If not head is defined, it will fallback to the default one (`<title>Wheelmap.org</title>`). That should be sufficient, no?

mutaphysis replied:
Sounds okay to me :)

mutaphysis commented on src/app/PlaceDetailsData.js:
```diff
@@ -147,27 +154,69 @@ const PlaceDetailsData: DataTableEntry<PlaceDetailsProps> = {
         }
+
+        const coordinates = normalizedCoordinatesForFeature(feature);
+        const thisPlaceIsOn = t`This place is on ${productName}: ${placeTitle}`;
```
-1: Add translator comment

mutaphysis commented on src/app/SearchData.js:
```diff
@@ -109,14 +108,12 @@ const SearchData: DataTableEntry<SearchProps> = {
+      <title key="title">
+        {getProductTitle(
+          clientSideConfiguration,
+          `Search results${searchQuery ? ` for “${searchQuery}”` : ''}`
```
-2: Untranslated, new string

mutaphysis commented on src/app/getInitialProps.js:
```diff
@@ -108,14 +113,21 @@ export async function getAppInitialProps(
     ? clientCache.translations
     : loadExistingLocalizationByPreference(locales);
 
+  const preferredLocale = translations[0].headers.language;
```
0: Maybe we want to give access to this as a return value somehow, unrelated to this PR.

mutaphysis commented on src/app/getInitialProps.js:
```diff
@@ -34,7 +34,10 @@ type DataTableQuery = {
-  getHead?: (props: Props & AppProps) => ?React$Element<Head>,
+  getHead?: (
+    props: Props & AppProps,
+    isMainApp: boolean
```
0: I am still confused what isMainApp does and why it is named like this

mutaphysis commented on src/lib/Feature.js:
```diff
@@ -466,3 +467,11 @@ export function removeNullAndUndefinedFields<T: {} | {}[]>(something: T): ?T {
   return something;
 }
+
+export function normalizedCoordinatesForFeature(feature: Feature): ?[number, number] {
```
👍

mutaphysis commented on src/lib/i18n.js:
```diff
@@ -100,6 +100,8 @@ function getTranslationsForLocale(locale): Translations {
   return translations[locale];
 }
 
+export const locales: string[] = Object.keys(translations);
```
0: Maybe availableLocales?

mutaphysis commented on src/pages/_app.js:
```diff
@@ -117,7 +135,13 @@ export default class App extends BaseApp {
+    // TODO Find a cleaner way to check if this is the main wheelmap app …
+    // Might not be needed anymore once the clientSideConfiguration contains information for
+    // Twitter and Google Analytics
+    const isMainApp = productName === 'Wheelmap';
```
0: Ah, now I see. Yes - this stuff should be part of the clientSideConfiguration at some point.

mutaphysis commented on src/pages/_document.js:
```diff
@@ -56,8 +67,19 @@ export default class MyDocument extends Document {
+          <meta content="follow index" name="robots" />
+
+          {/* Facebook app */}
+          <meta content="289221174426029" property="fb:app_id" />
```
-1: I guess this all should not be shown if !isMainApp

mutaphysis commented on src/pages/_app.js:
```diff
@@ -128,7 +152,32 @@ export default class App extends BaseApp {
+            {isMainApp && (
+              <GoogleAnalytics
+                siteVerification="ANQCVlgXNpnBdZInKyE4eWMeNPq-cIckJGMyEYPXNec"
+                trackingId="todo"
```
-2: Still missing the right tracking ID

lennerd commented:
Thanks for your review @mutaphysis! Added translations and translator comments and renamed `locales` to `availableLocales`. Do you like to take a second look?

mutaphysis commented on src/pages/_app.js:
```diff
@@ -117,24 +149,51 @@ export default class App extends BaseApp {
+            )}
+          </Head>
+          <OpenGraph productName={productName} description={description} />
+          {googleAnalytics && <GoogleAnalytics googleAnalytics={googleAnalytics} />}
```
👍

mutaphysis commented:
LGTM, great work
